### ISSUE TITLE
Relationship between DiffBuilder and JAXPXPathEngine

### ISSUE THREAD

**Issue by wiedehoeft** [xmlunit/xmlunit#255]
I have currently trouble because of very slow comparisons with the DiffBuilder class.
Changing the used xpath engine could be helpful but I am struggling with the relationship between DiffBuilder and JAXPXPathEngine.
Into the second class I can pass a custom XPathFactory (like Saxon), but I do not see a possibility to propagate this to the DiffBuilder.
Am I missing something? What is the relationship between the two classes, or is there just no relationship?
Then I would need a different approach.

Thanks for help,
I am currently a little bit confused.

**Comment by bodewig** [comment]
well, the `DifferenceEngine` itself doesn't use XPath at all, it just walks the DOM tree - and this is slow if your document is big.

If - and only if - you use something like `ElementSelectors.byXpath` then the XPath engine will be used. And right now this is not configurable at all: https://github.com/xmlunit/xmlunit/blob/main/xmlunit-core/src/main/java/org/xmlunit/diff/ElementSelectors.java#L401

**Comment by wiedehoeft** [comment]
The things I know so far:
1) The comparisons gets stucked in Diffbuilder.build(). The comparison needs sometimes hours for one document.
2) Because of this the xmlunit-library was cloned into the project and at every position where the `JAXPXPathEngine` was used the constructor was changed to:
`new JAXPXPathEngine(new net.sf.saxon.xpath.XPathFactoryImpl.XPathFactoryImpl())`. This improved performances drastically; from hours to minutes.

For using the official xmlunit version again one solution could be to make the `JAXPXPathEngine` configurable?

Or the DiffBuilder is misused at this place and a complete new solution should be found?

**Comment by bodewig** [comment]
If you can tell me why/where your `DiffBuilder` uses `new JAXPXPathEngine()` at all - so far I can only guess it is `ElementSelectors.byXpath` you use - I can answer your questions :-)

Maybe you can provide a snippet of your code that creates and sets up the `DiffBuilder` masking element names and other stuff that may be sensitive.

If you are using `ElementSelectors.byXpath` then we can easily add an overload that allows an `XPathFactory`  as additional argument and you'd be able to configure it.

Of course it is possible you are using `byXpath` even though there'd be an easier solution. `ElementSelector` is an interface that can be implemented via Java which may allow things to be sped up quite a bit at the price of writing some code.

**Comment by bodewig** [comment]
`GroupMatchingElementSelector` most likely is what I'd need to see (it is code not provided by XMLUnit) - this one probably uses the `XPathEngine` in some way. If your project forks XMLUnit anyway it would probably be easier to write that against against the XPath engine you want to use directly.

In your code I do not see any usage of any Hamcrest matcher at all, so I'm not sure how/where `EvaluateXPathMatcher` or `HasXPathMatcher` are involved. Both of them contain `withXPathFactory` methods that would allow you to use a custom `XPathFactory` in order to use Saxon right out of the box.

The `JAXPXPathEngine` code you show is not very different from the one in XMLUnit. I must admit I don't understand how `if (nodes instanceof NodeList)` can ever be false, though.

I will add new overloads to `ElementSelectors.byXpath` that accept an `XPathEngine` parameter as this is clearly missing right now. Without knowing what `GroupMatchingElementSelector` may look like I can only speculate. It may be quite possible that you'd get a bigger performance boost by not using XPath at all rather than by switching the XPath implementation, though.

**Comment by bodewig** [comment]
> I will add new overloads to ElementSelectors.byXpath

added with a6bb601

**Comment by bodewig** [comment]
the changed `byXpath` signatures are now available as SNAPSHOTs. Unless I get any bad feedback I'm going to cut a new release early next year and consider this issue done.

**Comment by wiedehoeft** [comment]
> the changed `byXpath` signatures are now available as SNAPSHOTs. Unless I get any bad feedback I'm going to cut a new release early next year and consider this issue done.

I would like to test it, but it will be a litte more complicated to use a snapshot version. I will check if I find a possibility in the company I am currently working for and give feedback during next week.

Thank you for fast answers and solutions!

**Comment by wiedehoeft** [comment]
I was able to test the new SNAPSHOT-version and the result was great!

I passed the saxon XPathFactoryImpl to ElementSelector.byXPath(...) and a customized version of the JAXPXPathEngine:
```
import javax.xml.transform.Source;
import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathExpressionException;
import javax.xml.xpath.XPathFactory;

import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import org.xmlunit.ConfigurationException;
import org.xmlunit.XMLUnitException;
import org.xmlunit.util.Convert;
import org.xmlunit.util.IterableNodeList;
import org.xmlunit.xpath.JAXPXPathEngine;

/**
* Override two methods, for performance improvements.
*/
public class CustomJAXPXPathEngine extends JAXPXPathEngine {

     private final XPath xpath;

     public CustomJAXPXPathEngine(final XPathFactory fac) {
         try {
              xpath = fac.newXPath();
         } catch (final Exception e) {
              throw new ConfigurationException(e);
         }
     }

     @Override
     public Iterable<Node> selectNodes(final String xPath, final Source s) {
         try {
              final Object nodes = xpath.evaluate(xPath, Convert.toInputSource(s),
                        XPathConstants.NODESET);
              if (nodes instanceof NodeList) {
                   return new IterableNodeList((NodeList) nodes);
              }

              if (nodes instanceof Iterable) {
                   return (Iterable<Node>) nodes;
              }
              return null;
         } catch (final XPathExpressionException ex) {
              throw new XMLUnitException(ex);
         }
     }

     @Override
     public Iterable<Node> selectNodes(final String xPath, final Node n) {
         try {
              final Object nodes = xpath.evaluate(xPath, n, XPathConstants.NODESET);

              if (nodes instanceof NodeList) {
                   return new IterableNodeList((NodeList) nodes);
              }

              if (nodes instanceof Iterable) {
                   return (Iterable) nodes;
              }

              return null;

         } catch (final XPathExpressionException ex) {
              throw new XMLUnitException(ex);
         }
     }
}
```
Now it is as fast as our customized version and with the new release we are able to use xmlunit as is again :-)

**Comment by bodewig** [comment]
Thank you for verifying. I've just published 2.9.1.
